### ISSUE TITLE
Add rules that "eslint-plugin-clean-regex" has

### ISSUE THREAD

**Issue by ota-meshi** [ota-meshi/eslint-plugin-regexp#91]
Check each rule that [eslint-plugin-clean-regex](https://github.com/RunDevelopment/eslint-plugin-clean-regex) has, add the missing rule, and if eslint-plugin-regexp already has a similar rule, check for missing validation is needed.
Also, if we keep a note of the rule correspondence, it may be a document when creating a migration tool.

If we want to discuss one rule, we can also open a new issue.

### Problems

- [x] [clean-regex/confusing-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/confusing-quantifier.md) ... Warn about confusing quantifiers.  
  - #93 Add new rule `regexp/confusing-quantifier`.
- [x] [clean-regex/disjoint-alternatives](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/disjoint-alternatives.md) ... Disallow different alternatives that can match the same words.
  - Handled by #179. 
- [x] [clean-regex/no-empty-alternative](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-empty-alternative.md) ... Disallow alternatives without elements.
  - #95 Add new rule `regexp/no-empty-alternative`.
- [x] [clean-regex/no-empty-backreference](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-empty-backreference.md) ... Disallow backreferences that will always be replaced with the empty string.
  - #111 makes `no-useless-backreference` functionally equivalent to `clean-regex/no-empty-backreference`.
- [x] [clean-regex/no-empty-lookaround](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-empty-lookaround.md) ... Disallow lookarounds that can match the empty string.
  - Similar to `regexp/no-empty-lookarounds-assertion`, but it doesn't catch potential empty assertions. `regexp/no-empty-lookarounds-assertion` needs to be changed. (#113)
- [x] [clean-regex/no-lazy-ends](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-lazy-ends.md) ... Disallow lazy quantifiers at the end of an expression.
  - #96 Add new rule `regexp/no-lazy-ends`.
- [x] [clean-regex/no-obscure-range](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-obscure-range.md) ... Disallow obscure ranges in character classes.
  - #122 adds this rule.
- [x] [clean-regex/no-octal-escape](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-octal-escape.md) ... Disallow octal escapes outside of character classes.
  - Similar to `regexp/no-octal`, but with some differences. We probably need to add options.
  - #108 makes `regexp/no-octal` even stricter than `clean-regex/no-octal-escape`.
- [x] [clean-regex/no-optional-assertion](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-optional-assertion.md) ... Disallow optional assertions.
  - #135 Add new rule `regexp/no-optional-assertion`.
- [x] [clean-regex/no-potentially-empty-backreference](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-potentially-empty-backreference.md) ... Disallow backreferences that reference a group that might not be matched.
  - #136 Add new rule `regexp/no-potentially-useless-backreference`.
- [x] [clean-regex/no-unnecessary-assertions](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-assertions.md) ... Disallow assertions that are known to always accept (or reject).
  - #137 Add new rule `regexp/no-useless-assertions`.
- [x] [clean-regex/no-zero-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-zero-quantifier.md) ... Disallow quantifiers with a maximum of 0.
  - ~~Can check with the `regexp/no-useless-exactly-quantifier` rule.~~
  - #199 & #200
- [x] [clean-regex/optimal-lookaround-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/optimal-lookaround-quantifier.md) ... Disallows the alternatives of lookarounds that end with a non-constant quantifier.
  - #97 Add new rule `regexp/optimal-lookaround-quantifier`.

### Suggestions

- [x] [clean-regex/consistent-match-all-characters](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/consistent-match-all-characters.md) ... Use one character class consistently whenever all characters have to be matched.
    - #103 handles some of it but we still need to add some options.
- [x] [clean-regex/identity-escape](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/identity-escape.md) ... How to handle identity escapes.
   - #220 `regexp/strict` rule instead
- [x] [clean-regex/no-constant-capturing-group](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-constant-capturing-group.md) ... Disallow capturing groups that can match only one word.
  - #188 brings over half of the rule.
- [x] [clean-regex/no-trivially-nested-lookaround](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-trivially-nested-lookaround.md) ... Disallow lookarounds that only contain another assertion.
  - #120 adds this rule.
- [x] [clean-regex/no-trivially-nested-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-trivially-nested-quantifier.md) ... Disallow nested quantifiers that can be rewritten as one quantifier.
  - #146 adds this rule. 
- [x] [clean-regex/no-unnecessary-character-class](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-character-class.md) ... Disallow unnecessary character classes.
  - Already fully covered by `regexp/no-useless-character-class`. 
- [x] [clean-regex/no-unnecessary-flag](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-flag.md) ... Disallow unnecessary regex flags.
  - #139 Add new rule `regexp/no-useless-flag`.
- [x] [clean-regex/no-unnecessary-group](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-group.md) ... Disallow unnecessary non-capturing groups.
  - #154 adds most of the functionality of `clean-regex/no-unnecessary-group` but it doesn't handle empty groups.
- [x] [clean-regex/no-unnecessary-lazy](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-lazy.md) ... Disallow unnecessarily lazy quantifiers.
  - #143 
- [x] [clean-regex/no-unnecessary-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/no-unnecessary-quantifier.md) ... Disallow unnecessary quantifiers.
  - #197 adds this rule. 
- [x] [clean-regex/optimal-concatenation-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/optimal-concatenation-quantifier.md) ... Use optimal quantifiers for concatenated quantified characters.
  - Added by #207. 
- [x] [clean-regex/optimized-character-class](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/optimized-character-class.md) ... Disallows unnecessary elements in character classes.
  - Mostly covered by `no-dupe-characters-character-class` after #191. 
  - The rest is already covered by `no-useless-range` and `prefer-range`.
- [x] [clean-regex/prefer-character-class](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/prefer-character-class.md) ... Prefer character classes wherever possible instead of alternations.
  - #158 updates `regexp/prefer-character-class`.
- [x] [clean-regex/prefer-predefined-assertion](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/prefer-predefined-assertion.md) ... Prefer predefined assertions over equivalent lookarounds.
  - #171 adds this rule. 
- [x] [clean-regex/prefer-predefined-character-set](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/prefer-predefined-character-set.md) ... Prefer predefined character sets instead of their more verbose form.
  - Already covered by `regexp/prefer-w` and `regexp/prefer-d`. 
- [x] [clean-regex/prefer-predefined-quantifiers](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/prefer-predefined-quantifiers.md) ... Prefer predefined quantifiers (+*?) instead of their more verbose form.
  - This is already covered by `regexp/prefer-question-quantifier`, `regexp/prefer-star-quantifier`, and `regexp/prefer-plus-quantifier`.
- [x] [clean-regex/simple-constant-quantifier](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/simple-constant-quantifier.md) ... Prefer simple constant quantifiers over the range form.
  - #109 makes `no-useless-two-nums-quantifier` functionally equivalent to `clean-regex/simple-constant-quantifier`.
- [x] [clean-regex/sort-flags](https://github.com/RunDevelopment/eslint-plugin-clean-regex/blob/master/docs/rules/sort-flags.md) ... Requires the regex flags to be sorted.
  - #164 Add new rule `regexp/sort-flags`.

**Comment by RunDevelopment** [comment]
Note: I just found a bug in `no-trivially-nested-lookaround` (RunDevelopment/eslint-plugin-clean-regex#30). Whoever brings the rule over should fix the bug.

**Comment by RunDevelopment** [comment]
Regarding `clean-regex/no-unnecessary-quantifier`: 
The rule is mostly covered by `no-useless-exactly-quantifier` but also does some other stuff that should probably be included in `confusing-quantifier`. I will need to think about that for a bit.

**Comment by ota-meshi** [comment]
I think we possible to mark no-useless-exactly-quantifier as deprecated and add new rule 😉

**Comment by RunDevelopment** [comment]
Regarding `clean-regex/disjoint-alternatives`:
Bringing this rule over will be tricky. It overlaps with `no-dupe-disjunctions` but it's not better than `no-dupe-disjunctions`. `disjoint-alternatives` is implemented using NFAs. This means that it can find non-trivial overlaps between alternatives that ``no-dupe-disjunctions`` can't find but it has 2 downside:

1. NFAs are expensive. While NFA operations are fairly optimized, they are not free.
2. Not all JS regexes can be converted into NFAs. I.e. `disjoint-alternatives` can't detect that `^a$` and `^a$` are the same. Assertions and backreferences can't be represented in NFAs so regexes containing them are difficult to deal with.

A hybrid approach using the current implementations of `no-dupe-disjunctions` and `disjoint-alternatives` has to be implemented.

Also want to talk about prefixes. `no-dupe-disjunctions` detects mistakes like `a|abc` (I do not understand why `disallowNeverMatch` is disabled by default). This is very useful but it's probably not complete (I say "probably" because the detection is so good that I couldn't find a simple example). Using NFAs, we can further improve the detection but it will be computationally expensive.

If it's ok with you @ota-meshi, then I'll bring over `disjoint-alternatives` some time this week

**Comment by ota-meshi** [comment]
> A hybrid approach using the current implementations of `no-dupe-disjunctions` and `disjoint-alternatives` has to be implemented.

I think this idea is good!

> (I do not understand why `disallowNeverMatch` is disabled by default)

Since `disallowNeverMatch` verifies patterns one by one, some patterns take a lot of time, so I turned off the default. The default behavior is that some patterns can be compared as strings, so I think it's fast in most cases.

> This is very useful but it's probably not complete

I think so, too. As far as I know, I haven't verified the `\p{}` and `i` flag (This can be improved using `regexp-ast-analysis`).
Other than that, it should work as far as I can think of, but I'm still not as familiar with regexp as you 😅.

**Comment by RunDevelopment** [comment]
> [...] some patterns take a lot of time, so I turned off the default. The default behavior is that some patterns can be compared as strings, so I think it's fast in most cases.

I ran the rule on Prism's 2k regexes this is the result:

| config | time |
| --- | --- |
| `'regexp/no-dupe-disjunctions': 'off'` | 2.9s |
| `'regexp/no-dupe-disjunctions': 'warn'` | 3.7s |
| `'regexp/no-dupe-disjunctions': ['warn', { disallowNeverMatch: true }]` | 5.4s |

And then there's `disjoint-alternatives`:

| config | time |
| --- | --- |
| `'clean-regex/disjoint-alternatives': 'off'` | 2.9s |
| `'clean-regex/disjoint-alternatives': 'warn'` | 5.6s |

Since `disjoint-alternatives` doesn't even implement a `disallowNeverMatch: true` option, the final hybrid approach will be even slower.

That being said, I don't think that this is a huge problem. Prism has literally thousands of complex regexes, the average code base will not see a noticeable performance impact. ESLint also provides options to cache results and to only run on changed files. 

We should also consider that `no-dupe-disjunctions` and `disjoint-alternatives` find real bugs that are hard to notice. These rules provide real value.

I agree that users must have the option to turn off advanced analysis if performance was a problem. But we shouldn't hide some of our best features by default.

> Other than that, it should work as far as I can think of [...].

Implement `disallowNeverMatch` for two alternatives `A|B`, we need to find out whether `B` is a subset of `A[^]*` (= `A` followed by any string).

Don't get me wrong, the `isCovered` function is really good but there will be some complex cases that it simply can't detect without doing a proper subset check. 
(That being said, the proper subset check is only possible in O(2^n). So we also won't be able to implement the subset check perfectly.)

**Comment by RunDevelopment** [comment]
Regarding `clean-regex/no-constant-capturing-group`: 
With #188, we support the default configuration of `clean-regex/no-constant-capturing-group`. I think we don't need to bring over the rest.

There are legitimate use cases for capturing groups that only match a single string so I don't want to just disallow them.

**Comment by ota-meshi** [comment]
@RunDevelopment 
I'm glad that most of the rule migrations are done :laughing:

The `clean-regex/identity-escape` rule is similar to the `regexp/no-useless-escape` rule, but the `clean-regex/identity-escape` rule seems to give user the flexibility to change the checks.

Do you think we can merge the features of the `clean-regex/identity-escape` rule by adding an option to the` regexp/no-useless-escape` rule? Or do you think we need to add new rule?

**Comment by RunDevelopment** [comment]
I'm not sure about `clean-regex/identity-escape`. 

The problem is that the rule can also add useless escaping (i.e. you can configure it to escape all `!` characters) which conflicts with ESLint's [`no-useless-escape` rule](https://eslint.org/docs/rules/no-useless-escape) and our `regexp/no-useless-escape` rule. 

The reason I added `clean-regex/identity-escape` is that I needed something like `regexp/no-useless-escape` but I also wanted something that enforces the use of `\{`, `\}`, and `\]` outside of character classes. We don't need an extra rule just for these 3 characters and merging this into `regexp/no-useless-escape` is weird.

Maybe we could use the fact that Unicode parsing mode basically has the rules for escaping that I want? We could add a new rule `regexp/strict` that makes sure any regex without the `u` flag is also syntactically valid with the `u` flag. This rule would have to add/remove escaping in exactly the way that I want in order to achieve its goal. `regexp/strict` will slightly overlap with `regexp/no-useless-escape` but not completely (e.g. `regexp/strict` will allow `/[\.]/`).

With `regexp/strict`, we could make my taste for escaping into something that is actually useful. This also gets rid of conflicts with the no-useless-escape rules. 

(It even gets rid of the overly complex rule option of `clean-regex/identity-escape` 😅 )

**Comment by ota-meshi** [comment]
The `regexp/strict` rule sounds good to me!

Maybe we can use `RegExpValidator`'s `strict` option.
https://github.com/mysticatea/regexpp/blob/master/src/validator.ts#L127

**Comment by RunDevelopment** [comment]
> Maybe we can use `RegExpValidator`'s `strict` option.

That's where I got the idea for the name from :)

**Comment by RunDevelopment** [comment]
With `regexp/strict` now being merged, I think we are done.

Thank you very much for your help @ota-meshi!
